docs/source/meta_init.rst

```diff
@@ -0,0 +1,239 @@
+    >>>
+    >>> # Do some form of inspection (e.g. apply a sharding algorithm).
+    >>>
+    >>> # Only materialize `sublayer1` and `sublayer2`.
+    >>> torch.nn.utils.materialize(m.sublayer1)
+    >>> torch.nn.utils.materialize(m.sublayer2)
+
+
+Note that :func:`meta_init` overrides even explicitly passed ``device``
+arguments and forces the use of the meta device.
+
+::
+
+    >>> class MyModule(torch.nn.Module):
+    ...     def __init__(self):
+    ...         super().__init__()
+    ...         self.param = torch.nn.Parameter(torch.ones([3], device="cpu"))
+    ...
+    >>> with torch.nn.utils.meta_init():
+    ...     m = MyModule()
+    >>> m.param
+    Parameter containing:
+    tensor(..., device='meta', size=(10, 10), requires_grad=True)
+    >>>
+    >>> torch.nn.utils.materialize(m)
+    >>> m.param
+    Parameter containing:
+    tensor([1., 1., 1.], requires_grad=True)
+
+
+Although the main use case of :func:`meta_init` is large module materialization,
+its underlying machinery is actually more flexible. You can also construct and
+later materialize individual tensors.
+
+::
+
+    >>> with torch.nn.utils.meta_init():
+    ...     t1 = torch.ones([3])
+    ...     t2 = t1 + 2
+    >>> t1
+    tensor(..., device='meta', size=(3,))
+    >>> t2
+    tensor(..., device='meta', size=(3,))
+    >>> # By default, `materialize()` clears the meta-init cache. If you want
+    ... # to materialize more than one object, set the `keep_cache` parameter
+    ... # to `True`.
+    >>> torch.nn.utils.materialize(t1, keep_cache=True)
+    >>> t1
+    tensor([1., 1., 1.])
+    >>> torch.nn.utils.materialize(t2)
+    >>> t2
+    tensor([3., 3., 3.])
+
+
+API
+---
+.. autosummary::
+    :toctree: generated
+    :nosignatures:
+
+    meta_init
+    is_meta_init
+    materialize
+    clear_meta_init_cache
+
+.. autofunction:: meta_init
+.. autofunction:: is_meta_init
+.. autofunction:: materialize
+.. autofunction:: clear_meta_init_cache
+
+.. _common-failures:
+
+Common Failure Cases
+--------------------
+Since meta initialization relies on constructing modules and their tensors on
+the meta device, there are some certain usage patterns that will fail when used
+with :func:`meta_init`.
+
+**A module that uses an operator not supported by the meta backend:**
+If the module uses an operator that is not yet supported by the meta backend,
+the operator call will fail. If you experience such failure, please open a
+GitHub issue and let us know which operator you are using.
+
+**A module that has device-specific initialization logic:**
+A module that expects to be initialized using one or more specific device types
+and includes initialization logic such as:
+
+::
+
+  def MyModule(Module):
+      def __init__(self):
+          x = torch.ones([10, 10])
+          if x.is_cpu():
+              y = x + 1
+          elif x.is_cuda():
+              y = x + 2
+
+
+will silently fail, potentially with some partial initialization.
```

Comment:
Probably need additional details here in terms of why it will fail and which lines causes issues and unexpected results.